module/profiler/profiler.go

```diff
@@ -143,11 +146,99 @@ func newProfileFunc(name string) profileFunc {
 	}
 }
 
-func (p *AutoProfiler) pprofAllocs(w io.Writer) error {
+func (p *AutoProfiler) goHeapProfile(sampleTypes ...string) (*profile.Profile, error) {
+	if len(sampleTypes) == 0 {
+		return nil, fmt.Errorf("no sample types specified")
+	}
+
 	// Forces the GC before taking each of the heap profiles and improves the profile accuracy.
 	// Autoprofiler runs very infrequently so performance impact is minimal.
 	runtime.GC()
-	return newProfileFunc("allocs")(w)
+
+	buf := &bytes.Buffer{}
+	err := newProfileFunc("heap")(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate allocs profile: %w", err)
+	}
+
+	prof, err := profile.Parse(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse allocs profile: %w", err)
+	}
+	prof.TimeNanos = time.Now().UnixNano()
+
+	// TODO(rbtz): add tests.
+	selectedSampleTypes := make([]int, 0, len(sampleTypes))
+	for _, name := range sampleTypes {
+		for i, sampleType := range prof.SampleType {
+			if sampleType.Type == name {
+				selectedSampleTypes = append(selectedSampleTypes, i)
+			}
+		}
+	}
+	if len(selectedSampleTypes) != len(sampleTypes) {
+		return nil, fmt.Errorf("failed to find all sample types: want: %+v, got: %+v", sampleTypes, prof.SampleType)
+	}
+
+	for i, j := range selectedSampleTypes {
+		prof.SampleType[i] = prof.SampleType[j]
```

Comment:
This is wrong.  you need to allocate a new slice (same below).  it accidentally works right now cuz you're passing in the type string argument in the correct/increasing order

Reply:
I've purposely did it inplace, but I would assume this indeed looks scary.  Let me re-create the slice instead.

Reply:
as i've pointed out, you did inline shifting incorrectly.  it works cuz the provided sampleTypes is in the correct order

you want something like this instead:

```
shiftIdx := 0
for i, sampleType := range prof.SampleType {
  include := false
  for _, name := range sampleTypes {
    if sampleType.Type == name {
      include = true
      break
    }
  }
  if include {
    prof.SampleType[shiftIdx] = sampleType
    shiftIdx++
  }
}
prof.SampleType = prof.SampleType[:shiftIdx]
```